### PR TITLE
vlan: enable ping test with vlan tag on windows guest

### PR DESCRIPTION
**Pull request by sitoliu**
id: 1579659
Signed-off-by: Sitong Liu <siliu@redhat.com>

### CODE REVIEW

**Comment by sitoliu** [comment]
JOB LOG    : /root/avocado/job-results/job-2018-05-31T16.41-d1dc251/job.log
 (1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.vlan: PASS (634.55 s)

File: generic/tests/cfg/vlan.cfg [hunk 1]
```diff
@@ -18,7 +17,12 @@
     RHEL.5:
         cmd_type = vconfig
         msg_pattern ='ERROR: trying to add VLAN #4095 to IF -:\w+:-  error: Invalid argument'
+    Windows:
+        vlan_num = 1
+        win_vlan_id = 900
+        set_vlan_cmd = powershell -command "Set-NetAdapter -Name '%s' -VlanID ${win_vlan_id} -Confirm:$false"
     variants:
         - @vlan_connective_test:
```
Comment (xiywang) [inline generic/tests/cfg/vlan.cfg]: is this also only for linux?

Comment (sitoliu) [reply generic/tests/cfg/vlan.cfg]: Thanks for review. 
No, I want to do ping test for both windows and linux, windows guest can do ping test on this variant.

File: generic/tests/vlan.py [hunk 2]
```diff
@@ -155,6 +155,20 @@ def flood_ping(src, dst):
...
+            set_vlan_cmd = params["set_vlan_cmd"]
+            nicid = utils_net.get_windows_nic_attribute(session=session,
+                                                        key="netenabled", value=True,
+                                                        target="netconnectionID")
+            session.cmd(set_vlan_cmd % nicid)
+            ifname.append(nicid)
+            dev_mac = vm.virtnet[0].mac
+            vm_ip.append(utils_net.get_guest_ip_addr(session, dev_mac, os_type="windows"))
+            session_ctl.append(session)
+            continue
```
Comment (hereischen) [inline generic/tests/vlan.py]: ~~`continue`~~

File: generic/tests/vlan.py [hunk 3]
```diff
@@ -188,6 +202,24 @@ def flood_ping(src, dst):
...
+        for vm_index, vm in enumerate(vms):
+            status, output = utils_test.ping(dest=vm_ip[(vm_index + 1) % 2], count=10,
+                                             session=session_ctl[vm_index],
+                                             timeout=30)
+            loss = utils_test.get_loss_ratio(output)
+            if not loss and re.findall(r"TTL=", output):
+                pass
+            else:
+                err_msg = "Guests ping test hit unexpected loss, "
+                err_msg = "error info: %s" % output
```
Comment (hereischen) [inline generic/tests/vlan.py]: Not correct here. I think you can just put the msg into test.fail()

File: generic/tests/vlan.py [hunk 4]
```diff
@@ -188,6 +202,24 @@ def flood_ping(src, dst):
             set_ip_vlan(session, vlan_i, v_ip, ifname[vm_index])
         set_arp_ignore(session)
 
+    if params["os_type"] == "windows":
+        for vm_index, vm in enumerate(vms):
+            status, output = utils_test.ping(dest=vm_ip[(vm_index + 1) % 2], count=10,
+                                             session=session_ctl[vm_index],
+                                             timeout=30)
+            loss = utils_test.get_loss_ratio(output)
+            if not loss and re.findall(r"TTL=", output):
```
Comment (hereischen) [inline generic/tests/vlan.py]: use `in` would be enough:
```python
if not loss and ("TTL=" in output):
```

Comment (sitoliu) [reply generic/tests/vlan.py]: Thanks, modified.

File: generic/tests/vlan.py (hunk 2 above)
Comment (sitoliu) [reply generic/tests/vlan.py]: Windows guest didn't need to do following steps, so add `continue` here...

Comment (hereischen) [reply generic/tests/vlan.py]: sorry， my fault.

**Comment by sitoliu** [comment]
JOB LOG    : /root/avocado/job-results/job-2018-06-06T16.42-c2d8817/job.log
 (1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.vlan: PASS (699.04 s)

**Comment by hereischen** [comment]
Acked-by: Haotong Chen hachen@redhat.com

File: generic/tests/vlan.py [hunk 5]
```diff
@@ -188,6 +203,22 @@ def flood_ping(src, dst):
             set_ip_vlan(session, vlan_i, v_ip, ifname[vm_index])
         set_arp_ignore(session)
 
+    if params["os_type"] == "windows":
+        for vm_index, vm in enumerate(vms):
+            status, output = utils_test.ping(dest=vm_ip[(vm_index + 1) % 2], count=10,
+                                             session=session_ctl[vm_index],
+                                             timeout=30)
+            loss = utils_test.get_loss_ratio(output)
+            if not loss and ("TTL=" in output):
```
Comment (pezhang) [inline generic/tests/vlan.py]: @sitoliu Could you please explain why "TTL=" is needed to show in the output? Thanks

**Comment by pezhang** [comment]
Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.vlan: FAIL (402.09 s)

10:34:35 ERROR| FAIL 6-Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.vlan -> TestFail: Guests ping test hit unexpected loss, error info:
Pinging DESKTOP-NGU7FFJ [::1] with 32 bytes of data:
Reply from ::1: time<1ms
Reply from ::1: time<1ms
Reply from ::1: time<1ms
Reply from ::1: time<1ms
Reply from ::1: time<1ms
Reply from ::1: time<1ms
Reply from ::1: time<1ms
Reply from ::1: time<1ms
Reply from ::1: time<1ms
Reply from ::1: time<1ms

Ping statistics for ::1:
    Packets: Sent = 10, Received = 10, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 0ms, Average = 0ms

C:\Windows\system32>

**Comment by pezhang** [comment]
Regarding above results, looks like it's ping localhost, not IP address.

**Comment by sitoliu** [comment]
@pezhang Windows guest need more time to get new ip, will extend the sleep time.

**Comment by sitoliu** [comment]
Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.vlan: PASS (405.26 s)

**Comment by sitoliu** [comment]
@pezhang Updated the code, set `linklocal=True` to `get_guest_ip_addr()`, since guest will have a linklocal ip after set vlan tag.

Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.vlan: PASS (756.86 s)
Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.vlan: PASS (595.92 s)

File: generic/tests/vlan.py [hunk 6]
```diff
@@ -155,6 +155,23 @@ def flood_ping(src, dst):
...
+            nicid = utils_net.get_windows_nic_attribute(session=session,
+                                                        key="netenabled", value=True,
+                                                        target="netconnectionID")
+            session.cmd(set_vlan_cmd % nicid, timeout=240)
+            time.sleep(10)
+            ifname.append(nicid)
+            dev_mac = vm.virtnet[0].mac
+            vm_ip.append(utils_net.get_guest_ip_addr(session, dev_mac,
+                                                     os_type="windows",
+                                                     linklocal=True))
```
Comment (pezhang) [inline generic/tests/vlan.py]: @sitoliu Could it be possible to print the vm ip address in the debug log? It can help to analyze test case if any issue exists.

**Comment by pezhang** [comment]
(1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.vlan: PASS (493.53 s)

Looks good to me. 
Acked-by: Pei Zhang pezhang@redhat.com

File: generic/tests/vlan.py (hunk 6 above)
Comment (sitoliu) [reply generic/tests/vlan.py]: @pezhang Thanks, updated it.

**Comment by pezhang** [comment]
@vivianQizhu Thanks.

(1/1) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.vlan: PASS (319.26 s)

(1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.vlan: PASS (341.89 s)
